### PR TITLE
Add recent blog posts on homepage

### LINKED ISSUE #10: Show recent blog posts on homepage
Unfortunately, at the moment, it's not a trivial task to do that in Docusaurus, but here are some ideas:
 - https://github.com/facebook/docusaurus/discussions/6423
 - https://kgajera.com/blog/display-recent-blog-posts-on-home-page-with-docusaurus/

Need to check up with the Docusaurus team if there's now an easier way in the latest release.

### PR DESCRIPTION
**Pull request by anchalshivank**
- Implemented a custom blog plugin to generate recent-posts.json
- Integrated recent-posts.json data into RecentBlogs component
- Enhanced RecentBlog component to include title, date, reading time, and description
- Added CSS modules for styling RecentBlog and RecentBlogs components

### CODE REVIEW

File: src/pages/index.jsx [hunk 1]
```diff
@@ -36,13 +38,13 @@ function HomepageHeader() {
...
 export default function Home() {
   const { siteConfig } = useDocusaurusContext();
-
   useBrokenLinks().collectAnchor('#contact-email');
 
   return (
     <Layout
       description={siteConfig.customFields.description}>
       <HomepageHeader />
+      <RecentBlogs />
```
Comment (sergefdrv) [inline src/pages/index.jsx]: Let's move it to the bottom of the page and fit into the [Figma grid](https://infima.dev/docs/layout/grid) similar to section "What is Replica_IO about?".

File: plugins/blog-plugin.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Override the default `contentLoaded` hook to access blog posts data
+     */
+    contentLoaded: async function (params) {
+      const { content, actions } = params;
+
+      // Get the 5 latest blog posts
+      const recentPostsLimit = 5;
+      const recentPosts = content.blogPosts.slice(0, recentPostsLimit);
+      actions.createData('recent-posts.json',recentPosts);
```
Comment (sergefdrv) [inline plugins/blog-plugin.js]: Let's export the metadata of all posts here, but avoid dumping the content. We can choose how many to render in the recent blog posts component. The output file would be then named differently, e.g. `blog-posts-metadata.json`.

File: src/components/RecentBlogs/index.jsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import recentBlogs from '@site/.docusaurus/docusaurus-plugin-content-blog/default/recent-posts.json';
+import RecentBlog from '../RecentBlog';
+import styles from './index.module.css';
+
+const RecentBlogs = () => {
```
Comment (sergefdrv) [inline src/components/RecentBlogs/index.jsx]: Let's parameterize this component with the number of recent posts to show. The parameter should be supplied by the index page from the site configuration.

File: src/components/RecentBlog/index.module.css [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+.blogMeta {
+    display: block;
+    margin-top: 0.5rem;
+    color: #666;
+}
+
+.blogDescription {
+    margin-top: 0.75rem;
+    color: #444;
+}
```
Comment (sergefdrv) [inline src/components/RecentBlog/index.module.css]: This styling doesn't look good in the dark mode. We really need to use [Figma variables](https://github.com/facebookincubator/infima/blob/main/packages/core/styles/common/variables.pcss) for most of the properties like colors, shadows, font sizes, etc.

**Review by sergefdrv — CHANGES_REQUESTED** [review]

File: plugins/blog-plugin.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+          hasTruncateMarker: post.metadata.hasTruncateMarker,
+          authors: post.metadata.authors,
+          frontMatter: post.metadata.frontMatter,
+          unlisted: post.metadata.unlisted,
+          lastUpdatedAt: post.metadata.lastUpdatedAt,
+          lastUpdatedBy: post.metadata.lastUpdatedBy,
+          prevItem: post.metadata.prevItem
+        }
+      }));
+      actions.createData('blog-posts-metadata.json',recentPosts);
```
Comment (sergefdrv) [inline plugins/blog-plugin.js]:
```js
import { shouldBeListed } from '@docusaurus/plugin-content-blog/src/blogUtils';
```

```suggestion
      const posts = content.blogPosts
        .filter(shouldBeListed)
        .map(({ content: _, ...post }) => post);
      actions.createData('blog-posts-metadata.json', posts);
```

File: src/components/RecentBlog/index.jsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+            <a href={metadata.permalink} className={styles.blogTitle}>{metadata.title}</a>
+            <span className={styles.blogMeta}>
+                {formattedDate} · {readingTime} min read
+            </span>
+            <p className={styles.blogDescription}>{metadata.description}</p>
+        </div>
+    );
+};
+
+export default RecentBlog;
```
Comment (sergefdrv) [inline src/components/RecentBlog/index.jsx]: Let's rename it (and the file) to `RecentBlogPost`

File: src/components/RecentBlog/index.jsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    const readingTime = Math.round(metadata.readingTime);
+
+    return (
+        <div className={styles.blogItem}>
+            <a href={metadata.permalink} className={styles.blogTitle}>{metadata.title}</a>
+            <span className={styles.blogMeta}>
+                {formattedDate} · {readingTime} min read
+            </span>
+            <p className={styles.blogDescription}>{metadata.description}</p>
+        </div>
```
Comment (sergefdrv) [inline src/components/RecentBlog/index.jsx]: Let's try using [Figma cards](https://infima.dev/docs/components/card), similar to how Docusaurus website [renders](https://github.com/facebook/docusaurus/blob/main/website/src/components/Tweet/index.tsx#L33-L58) selected tweets [on its home page](https://github.com/facebook/docusaurus/blob/main/website/src/pages/index.tsx#L75-L99). We don't need to show avatars, only post's title, description, date, and reading time. I think we won't need much of the CSS if we use Figma cards. Would also be nice if the whole card becomes a clickable link to the post.

File: src/components/RecentBlogs/index.jsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        <div>
+            <h1>Recent Blog Posts</h1>
+            {blogs.map((blog, index) => (
+                <RecentBlog key={index} blog={blog} />
+            ))}
+        </div>
+    );
+};
+
+export default RecentBlogs;
```
Comment (sergefdrv) [inline src/components/RecentBlogs/index.jsx]: Let's rename it (and the file) to `RecentBlogPosts`

File: src/pages/index.jsx [hunk 9]
```diff
@@ -115,6 +117,9 @@ export default function Home() {
             </p>
           </p>
         </p>
+        <div className="container padding-vert--xl">
+          <RecentBlogs blogs_count={2} />
```
Comment (sergefdrv) [inline src/pages/index.jsx]: Let's take the number of recent posts to show from `siteConfig`. We can add `recentBlogPostsOnHomePage` to `customFields` in `docusaurus.config.js`.

File: src/components/RecentBlogs/index.jsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import recentBlogs from '@site/.docusaurus/docusaurus-plugin-content-blog/default/blog-posts-metadata.json';
```
Comment (sergefdrv) [inline src/components/RecentBlogs/index.jsx]:
```suggestion
import posts from '@site/.docusaurus/docusaurus-plugin-content-blog/default/blog-posts-metadata.json';
```

File: src/components/RecentBlogs/index.jsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import RecentBlog from '../RecentBlog';
+import styles from './index.module.css';
+
+const RecentBlogs = ({ blogs_count }) => {
+
+    const blogs = recentBlogs.slice(0, blogs_count);
```
Comment (sergefdrv) [inline src/components/RecentBlogs/index.jsx]:
```suggestion
    const recentPosts = posts.slice(0, blogs_count);
```

File: src/components/RecentBlog/index.jsx [hunk 12]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import styles from './index.module.css';
+
+const RecentBlog = ({ blog }) => {
```
Comment (sergefdrv) [inline src/components/RecentBlog/index.jsx]:
```suggestion
const RecentBlog = ({ post }) => {
```

File: plugins/blog-plugin.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
+const blogPluginExports = require('@docusaurus/plugin-content-blog');
+
+const defaultBlogPlugin = blogPluginExports.default;
+
+async function blogPluginExtended(...pluginArgs) {
+  const blogPluginInstance = await defaultBlogPlugin(...pluginArgs);
+
+  const pluginOptions = pluginArgs[1];
+
```
Comment (sergefdrv) [inline plugins/blog-plugin.js]:
```suggestion
```

**Review by sergefdrv — CHANGES_REQUESTED** [review]

File: src/components/RecentBlogPosts/index.jsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import posts from '@site/.docusaurus/docusaurus-plugin-content-blog/default/blog-posts-metadata.json';
+import RecentBlog from '../RecentBlogPost';
+
+const RecentBlogPosts = ({ blogs_count }) => {
+    const recentPosts = posts.slice(0, blogs_count);
+
+    if (recentPosts.length === 0) {
+        return <p>No recent blog posts available.</p>;
```
Comment (sergefdrv) [inline src/components/RecentBlogPosts/index.jsx]: I think it is better to return nothing in this case.

Comment (anchalshivank) [reply src/components/RecentBlogPosts/index.jsx]: will return null

File: src/components/RecentBlogPosts/index.jsx [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import posts from '@site/.docusaurus/docusaurus-plugin-content-blog/default/blog-posts-metadata.json';
+import RecentBlog from '../RecentBlogPost';
+
+const RecentBlogPosts = ({ blogs_count }) => {
+    const recentPosts = posts.slice(0, blogs_count);
```
Comment (sergefdrv) [inline src/components/RecentBlogPosts/index.jsx]:
```suggestion
const RecentBlogPosts = ({ nrPosts }) => {
    const recentPosts = posts.slice(0, nrPosts);
```

File: src/pages/index.jsx [hunk 16]
```diff
@@ -39,6 +41,8 @@ export default function Home() {
 
   useBrokenLinks().collectAnchor('#contact-email');
 
+  const blogsCount = siteConfig.customFields.recentBlogPostsOnHomePage;
```
Comment (sergefdrv) [inline src/pages/index.jsx]:
```suggestion
  const nrRecentBlogPosts = siteConfig.customFields.recentBlogPostsOnHomePage;
```

File: src/pages/index.jsx [hunk 17]
```diff
@@ -115,6 +119,9 @@ export default function Home() {
             </p>
           </p>
         </p>
+        <div className="container padding-vert--xl">
+          <RecentBlogs blogs_count={blogsCount} />
```
Comment (sergefdrv) [inline src/pages/index.jsx]:
```suggestion
          <RecentBlogs nrPosts={nrRecentBlogPosts} />
```

**Review by sergefdrv — CHANGES_REQUESTED** [review]
Thanks, save for a few nit-picking suggestions, this PR looks nice! Please apply the suggestions and squash the commits :+1:

File: src/pages/index.jsx [hunk 18]
```diff
@@ -6,6 +6,8 @@ import clsx from 'clsx';
 
 import Heading from '@theme/Heading';
 import styles from './index.module.css';
+import RecentBlogs from '../components/RecentBlogPosts';
```
Comment (sergefdrv) [inline src/pages/index.jsx]:
```suggestion
import RecentBlogPosts from '../components/RecentBlogPosts';
```

File: src/pages/index.jsx [hunk 19]
```diff
@@ -115,6 +119,9 @@ export default function Home() {
             </p>
           </p>
         </p>
+        <div className="container padding-vert--xl">
+          <RecentBlogs nrPosts={nrRecentBlogPosts} />
```
Comment (sergefdrv) [inline src/pages/index.jsx]:
```suggestion
          <RecentBlogPosts nrPosts={nrRecentBlogPosts} />
```

File: src/components/RecentBlogPosts/index.jsx [hunk 20]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import posts from '@site/.docusaurus/docusaurus-plugin-content-blog/default/blog-posts-metadata.json';
+import RecentBlog from '../RecentBlogPost';
```
Comment (sergefdrv) [inline src/components/RecentBlogPosts/index.jsx]:
```suggestion
import RecentBlogPost from '../RecentBlogPost';
```

File: src/components/RecentBlogPosts/index.jsx [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+    if (recentPosts.length === 0) {
+        return null;
+    }
+
+    return (
+        <div>
+            <h1>Recent Blog Posts</h1>
+            <div className="card__wrapper" style={{ display: 'grid', gap: '1rem' }}>
+                {recentPosts.map((post, index) => (
+                    <RecentBlog key={index} post={post} />
```
Comment (sergefdrv) [inline src/components/RecentBlogPosts/index.jsx]:
```suggestion
                    <RecentBlogPost key={index} post={post} />
```

**Review by sergefdrv — CHANGES_REQUESTED** [review]
Some final nit-picking. The commit message should not mention the history of the PR (like "Enhanced `RecentBlogPost` component"). Instead, it would better tell why we had to hack around the blog plugin. The rest seems to be quite self-explanatory.

File: src/components/RecentBlogPosts/index.jsx (hunk 20 above)
Comment (sergefdrv) [reply src/components/RecentBlogPosts/index.jsx]: Perhaps, `RecentBlogPost` could be included directly into `RecentBlogPosts/index.jsx`; up to you.

**Review by sergefdrv — APPROVED** [review]
Nice! Thanks :pray:

**Pull request merged**